File: backend-py/xiaoman/tools/time_sense.py
"""Time Sense Tool — 使用 TimeService（PRD memory-03 §7.3）"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from xiaoman.session import XiaomanSession
from xiaoman.time_service import TimeService
from xiaoman.tool_registry import XiaomanTool


class TimeSenseTool(XiaomanTool):
    """获取当前时间感知 — LLM 可调用"""

    @property
    def name(self) -> str:
        return "time_sense"

    @property
    def description(self) -> str:
        return "获取当前时间信息（时段、日期、星期、特殊提醒），用于时间感知回复。"

    @property
    def parameters(self) -> Mapping[str, Any]:
        return {"type": "object", "properties": {}}

    def __call__(self, session: XiaomanSession, arguments: Mapping[str, Any]) -> str:
        ts = TimeService()
        ctx = ts.get_time_context()
        weekdays = ["周一", "周二", "周三", "周四", "周五", "周六", "周日"]
        return (
            f"现在{ctx.day_type}{ctx.period}，"
            f"{ctx.hour}点，{weekdays[ctx.weekday]}，季节{ctx.season}。"
            f"小满状态：{ctx.xiaoman_status}。"
        )

File: backend/src/tools/time-sense.ts
export function createTimeSenseTool(): any {
  return {
    name: "time_sense",
    description: "获取当前时间，决定小满的开场白和状态",
    parameters: {
      type: "object",
      properties: {},
    },
    execute: async () => {
      const now = new Date();
      const hour = now.getHours();
      let greeting: string;
      let mood: string;

      if (hour >= 6 && hour < 11) {
        greeting = "早啊...我昨晚没睡好，一直在想那道物理题";
        mood = "困";
      } else if (hour >= 11 && hour < 14) {
        greeting = "食堂今天有糖醋排骨但我没抢到，你在干嘛？";
        mood = "懒";
      } else if (hour >= 14 && hour < 18) {
        greeting = "下午好烦啊，数学课我差点睡着";
        mood = "烦";
      } else if (hour >= 18 && hour < 22) {
        greeting = "终于下课了！我今天被英语老师点名了，救命";
        mood = "开心";
      } else {
        greeting = "这么晚了你还在这？我作业还没写完";
        mood = "焦虑";
      }

      return { time: now.toISOString(), hour, greeting, mood };
    },
  };
}
